### ISSUE TITLE
Should get_flag_value return an array?

### ISSUE THREAD

**Issue by paulray** [nanograv/PINT#581]
Hi, I wonder if `get_flag_value` should return an array. I  just used it to select all TOAs that met a criteria based on the value in a flag, like this:
```python
tt = pint.toa.get_TOAs('topobary.tim')

# Select the TOAs with the "htest" flag > 9.0
np.array(tt.get_flag_value('htest')) > 9.0                                                                                                       
tt.select(good)                              
```
If `get_flag_value` returned an array I wouldn't have to do the cast to `np.array`.
Thoughts?

**Comment by bwmeyers** [comment]
I would think that if it always returns a `list` at the moment anyway, then there should be no harm in returning a `numpy` array instead.

**Comment by paulray** [comment]
I don't think this is used anywhere except in the unit tests, so it should be safe to change.
I can imagine a flag that returns heterogeneous values, so determining the right type for the array might be ambiguous? But certainly for flags that are all numbers it makes sense to return them as an array.

**Comment by luojing1211** [comment]
This function will be used in the wideband TOAs for getting DM values from
the flags. But, it also used for getting strings, with different lengths
possibly, for instance, backend name.  So a type check is a must.
-Jing

**Comment by paulray** [comment]
I assume a numpy array of strings will be fine for backends, right?
But, if not, we could just add a keyword argument like `asarray=True` that would return it as an array if desired by the caller.

**Comment by scottransom** [comment]
Yeah, this is a bit tricky.  Because in general, the flags are not specified to be of any specific type and so at least as a first step we need to treat them all as strings.  In fact, I think I'd prefer that as the default.  And we can cast them to appropriate numbers (or whatever) as needed.  But we need to the user to specify what type that should be.

**Comment by luojing1211** [comment]
I think get_flag_value() returning strings is the way to go. There are
flags only apply to a subset of TOAs, and those flags values can give as
empty strings.

-Jing

**Comment by paulray** [comment]
Perhaps `get_flag_list()` and `get_flag_array()` could be there?  Then the array version could have some added smarts to make a masked array or something like that if there are TOAs with missing values?

**Comment by aarchiba** [comment]
The underlying problem is types of flags - if flags are convenient, it
would be really nice if floating-point values were actually floating-point
values (optionally NaN or masked) at which point we might as well be
returning arrays (which we can do for strings too). Of course, the tempo2
format "spec" doesn't provide a way for flags to specify types of their
values, or even whether they have values at all. But I think in the
interest of usability it would be a big help if PINT contained a list of
known flags, to which users could add, each of which specified a type of
their argument (float, int, string, bool/zero-or-one-int). Then reading tim
files could validate flag values as well, and python-level users could rely
on the flags having sensible types and coming in arrays.

Software-wise, the implementation is fairly easy. Policy-wise, PINT could
maintain a list of publicly known flags (so you'd know you didn't need to
introduce yet another alias for "band" or "receiver") and their types. As
we should for standardizing other aspects of par and tim files.

Anne

**Comment by paulray** [comment]
@luojing1211 now that you've done the wideband TOA stuff, any thoughts on this? Should we close it, or would it be good to add a method that returns flag values as an array?

**Comment by luojing1211** [comment]
Right now get_flag_values() returns a list of all the value and a list of valid value indices. Add a get_flag_value_as_array() is not a bad idea. I think Anne's idea is also interesting. I would rather to have a PR to address this issue an then close this one.
